### PR TITLE
chore: add active time to stream consumers

### PR DESCRIPTION
**Pull request by romange**
Adjust XINFO command to output active-time property. Store active-time and switch to RDB_TYPE_STREAM_LISTPACKS_3 if FLAGS_stream_rdb_encode_v2 is enabled.

### CODE REVIEW

File: src/server/rdb_save.cc [hunk 1]
```diff
@@ -215,8 +215,8 @@ uint8_t RdbObjectType(const PrimeValue& pv) {
       }
       break;
     case OBJ_STREAM:
-      return absl::GetFlag(FLAGS_stream_rdb_encode_v2) ? RDB_TYPE_STREAM_LISTPACKS
-                                                       : RDB_TYPE_STREAM_LISTPACKS_2;
+      return absl::GetFlag(FLAGS_stream_rdb_encode_v2) ? RDB_TYPE_STREAM_LISTPACKS_3
```
Comment (BorysTheDev) [inline src/server/rdb_save.cc:218]: looks strange to me

File: src/server/stream_family.cc [hunk 2]
```diff
@@ -82,7 +82,8 @@ struct NACKInfo {
 
 struct ConsumerInfo {
   string name;
-  size_t seen_time;
+  uint64_t seen_time;
+  int64_t active_time;
```
Comment (BorysTheDev) [inline src/server/stream_family.cc:86]: can we have the same type for time?

File: src/server/stream_family.cc [hunk 3]
```diff
@@ -1254,6 +1257,14 @@ OpResult<ClaimInfo> OpClaim(const OpArgs& op_args, string_view key, const ClaimO
 
   StreamMemTracker tracker;
 
+  // Try to get the consumer. If not found, create a new one.
+  auto cname = WrapSds(opts.consumer);
+  streamConsumer* consumer = streamLookupConsumer(cgr_res->cg, cname);
+  if (consumer == nullptr)
+    consumer = StreamCreateConsumer(cgr_res->cg, opts.consumer, now_ms, SCC_DEFAULT);
+  else
+    consumer->seen_time = now_ms;
```
Comment (BorysTheDev) [inline src/server/stream_family.cc:1266]: I want to suggest making a separate function UpdateOrCreateConsumer

File: src/server/rdb_save.cc (hunk 1 above)
Comment (romange) [reply src/server/rdb_save.cc:218]: I do not understand the comment

File: src/server/stream_family.cc (hunk 3 above)
Comment (romange) [reply src/server/stream_family.cc:1266]: sure

File: src/server/stream_family.cc [hunk 4]
```diff
@@ -3099,26 +3113,17 @@ variant<bool, facade::ErrorReply> HasEntries2(const OpArgs& op_args, string_view
...
-      consumer = StreamCreateConsumer(group, opts->consumer_name, op_args.db_cntx.time_now_ms,
-                                      SCC_NO_NOTIFY | SCC_NO_DIRTIFY);
-    }
+    sds cname = WrapSds(opts->consumer_name);
+    consumer = streamLookupConsumer(group, cname);
+    uint64_t now_ms = op_args.db_cntx.time_now_ms;
+    if (consumer)
+      consumer->seen_time = now_ms;
+    else
+      consumer = StreamCreateConsumer(group, opts->consumer_name, now_ms, SCC_DEFAULT);
```
Comment (BorysTheDev) [inline src/server/stream_family.cc:3122]: if you create a function UpdateOrCreateConsumer() you can use it here

File: src/server/stream_family.cc [hunk 5]
```diff
@@ -3130,6 +3135,14 @@ variant<bool, facade::ErrorReply> HasEntries2(const OpArgs& op_args, string_view
       requested_sitem.id.val = requested_sitem.group->last_id;
       streamIncrID(&requested_sitem.id.val);
     }
+  } else {
+    // Resolve $ to the last ID in the stream.
+    if (requested_sitem.id.resolve_last_id) {
+      requested_sitem.id.val = last_id;
+      streamIncrID(&requested_sitem.id.val);  // include id's strictly greater
```
Comment (BorysTheDev) [inline src/server/stream_family.cc:3142]: should we check the result of streamIncrID().

**Review by BorysTheDev — APPROVED** [review]

File: src/server/rdb_save.cc (hunk 1 above)
Comment (BorysTheDev) [reply src/server/rdb_save.cc:218]: You write it in the description that we will use RDB_TYPE_STREAM_LISTPACKS_3 for FLAGS_stream_rdb_encode_v2,
but it looks strange to me

File: src/server/stream_family.cc (hunk 5 above)
Comment (romange) [reply src/server/stream_family.cc:3142]: I backported the changes from valkey. They do not check it.

Comment (BorysTheDev) [reply src/server/stream_family.cc:3142]: we check it sometimes and sometimes no. We can rewrite this method in this case and never check for example

**Review by BorysTheDev — APPROVED** [review]

**Pull request merged**